kvserver: fully adopt useReproposalsV2

tbg opened the pull request:
This hard-codes useReproposalsV2 to true (prior to this PR, it's metamorphic with a default of true) and applies the resulting simplifications across the codebase.

These include simplified control flow and an overall simplified architecture.

Closes https://github.com/cockroachdb/cockroach/issues/105625.

Epic: CRDB-25287
Release note: None

tbg committed:
kvserver: hard-code useReproposalsV2 to true

tbg committed:
kvserver: rm retrieveLocalProposals(V1)

tbg committed:
kvserver: remove weird v1-proposals code in propose

So glad to see this go.

tbg committed:
kvserver: delete TestReplicaRefreshMultiple and TestReplicaReproposalWithNewLeaseIndexError

These tests don't work with v2 and aren't maintainable or super useful anyway.

We have good randomized coverage of these code paths now via kvnemesis and
TestProposalsWithInjectedLeaseIndexAndReproposalError, and adding better unit
test coverage is tracked in
https://github.com/cockroachdb/cockroach/issues/106504. In addition to this, it
would be desirable to extend this unit test coverage upwards through the stack
via https://github.com/cockroachdb/cockroach/issues/105177.

tbg committed:
kvserver: bake in useReproposalsV2 in prepareLocalResult

tbg committed:
kvserver: add a TODO

There are some invariants waiting to happen here.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106750)
<!-- Reviewable:end -->

tbg committed:
kvserver: rename retrieveLocalProposals{V2,}

The comments will be fixed up separately. This change is mechanical.

tbg committed:
kvserver: update comment on retrieveLocalProposals

tbg committed:
kvserver: rename in preparation for inlining

tbg committed:
kvserver: inline tryReproposeWithNewLeaseIndexShared

tbg committed:
kvserver: rename tryReproposeWithNewLeaseIndex{V2,}

tbg committed:
kvserver: brush up comments around ProposalData

The semantics are simpler now.

tbg committed:
kvserver: always enable assertion

useReproposalsV2 == true.

tbg committed:
kvserver: update endCmds.done

tbg committed:
kvserver: document expectations about single-release of latches

I can't think of reasons why endCmds.done (which releases latches, etc) should
be called twice. Document that expectation with a TODO to verify it.

tbg committed:
kvserver: remove (now) unused useReproposalsV2

tbg committed:
kvserver: update proposal lifecycle diagram

It's nice to see all of this complexity melt away.

tbg committed:
kvserver: improve a comment

erikgrinaker commented on pkg/kv/kvserver/replica_application_decoder.go:
```diff
@@ -75,15 +75,13 @@ func (d *replicaDecoder) decode(ctx context.Context, ents []raftpb.Entry) error
+// waiting on their result, and may be reproposed (as a new proposal) with a new
+// lease index in case they apply with an illegal lease index (see
+// tryReproposeWithNewLeaseIndex).
+func (d *replicaDecoder) retrieveLocalProposals() (anyLocal bool) {
```
nit: V2 rename should be in previous commit, or alternatively just squash the entire commit into it.

erikgrinaker commented on pkg/kv/kvserver/replica_proposal.go:
```diff
@@ -52,69 +52,59 @@ import (
+//	         └──apply goroutine◄───────┘
+//	                   │
+//	                   ▼
+//	                finished
```
❤️

erikgrinaker reviewed (approved):
LGTM, thanks for the cleanup!

tbg commented:
TFTR!

Comments addressed.

bors r=erikgrinaker

tbg commented:
TestRules/join_order flaked, pointed it out [here](https://cockroachlabs.slack.com/archives/C016CAD2HQ8/p1689743545636259)

bors r=erikgrinaker
